include/robot_interfaces/logger.hpp

```diff
@@ -0,0 +1,287 @@
+
+                        output_file_ << logger_data_->observation->timestamp_s(j) << " , " << j << " , " ;
+
+                        for(loop_iterator_ = 0; loop_iterator_ < observation_data_.size(); loop_iterator_++)
+                        {
+                        std::copy(observation_data_[loop_iterator_].begin(), observation_data_[loop_iterator_].end(), output_iterator_);
+                        }
+
+                        for(loop_iterator_ = 0; loop_iterator_ < applied_action_data_.size(); loop_iterator_++)
+                        {
+                          std::copy(applied_action_data_[loop_iterator_].begin(), applied_action_data_[loop_iterator_].end(), output_iterator_);
+                        }
+
+                        for(loop_iterator_ = 0; loop_iterator_ < desired_action_data_.size(); loop_iterator_++)
+                        {
+                          std::copy(desired_action_data_[loop_iterator_].begin(), desired_action_data_[loop_iterator_].end(), output_iterator_);
+                        }
+
+                        output_file_ << std::endl;
+
+                        std::cout << j << std::endl;
+
+                    }
+
+                    catch (const std::exception &e)
+                    {
+                        std::cout << "Trying to access index older than the "
+                                     "oldest! Skipping ahead."
+                                  << std::endl;
+                    }
+                  }
+
+                    index_ += block_size_;
+
+                    output_file_.close();
+
+// to check whether the data being requested to be logged is in the buffer of
+// the timeseries and inspect effect of delays
+#ifdef VERBOSE
+                    std::cout << "Index trying to access, oldest index in the "
+                                 "buffer: "
+                              << j << ","
+                              << logger_data_->observation->oldest_timeindex()
+                              << std::endl;
+#endif
+
+// to print the time taken for one block of data to be logged.
+#ifdef VERBOSE
+                auto t2 = std::chrono::high_resolution_clock::now();
+                auto duration =
+                    std::chrono::duration_cast<std::chrono::microseconds>(t2 -
+                                                                          t1)
+                        .count();
+
+                std::cout << "Time taken for one block of data to be logged: "
+                          << duration << std::endl;
+#endif
+
+            }
+        }
+    }
+
+    void start()
+    {
+        thread_->create_realtime_thread(&FingerLogger::write, this);
+    }
+
+    void stop()
+    {
+      output_file_.open(output_file_name_, std::ios_base::app);
+      std::ostream_iterator<double> output_iterator_(output_file_, " , ");
+
+      for(int j = index_; j < logger_data_->observation->newest_timeindex(); j++)
```

Comment:
There is a lot of redundancy here with the `write` function.  To avoid this, you could move the redundant part into a separate function that is called by both.
I think the difference is mainly the loop condition.  You could use something like `min(index_ + block_size_, newest_timeindex())` which should work for both cases.

Reply:
I didn't really get this and will ask you about it.